include/pybind11/stl_bind.h

```diff
@@ -397,14 +397,20 @@ vector_buffer(Class_& cl) {
         if (!detail::compare_buffer_info<T>::compare(info) || (ssize_t) sizeof(T) != info.itemsize)
             throw type_error("Format mismatch (Python: " + info.format + " C++: " + format_descriptor<T>::format() + ")");
 
-        auto vec = std::unique_ptr<Vector>(new Vector());
-        vec->reserve((size_t) info.shape[0]);
         T *p = static_cast<T*>(info.ptr);
         ssize_t step = info.strides[0] / static_cast<ssize_t>(sizeof(T));
         T *end = p + info.shape[0] * step;
-        for (; p != end; p += step)
-            vec->push_back(*p);
-        return vec.release();
+        if (step == 1) {
+            auto vec = std::unique_ptr<Vector>(new Vector(p, end));
+            return vec.release();
+        }
+        else {
+            auto vec = std::unique_ptr<Vector>(new Vector());
+            vec->reserve((size_t) info.shape[0]);
+            for (; p != end; p += step)
+                vec->push_back(*p);
+            return vec.release();
+        }
```

Comment:
Out of curiosity: can we get rid of `std::unqiue_ptr`? I'm not aware of any performance penalties of return-by-value alternative which looks much cleaner. 

```suggestion
        if (step == 1) {
            return Vector(p, end);
        }
        else {
            Vector vec((size_t) info.shape[0]);
            for (; p != end; p += step)
                vec.push_back(*p);
            return vec;
        }
```

Reply:
I don't see any issues with doing that, performance looks the same to me.

Reply:
It definitely looks cleaner, but there's a possibility that only one of the two branches will have return value optimization applied. In theory, it should be possible to apply RVO to both branches, but I don't think compilers do that (yet). Can we check what are the compilers actually doing?

I mean, we're getting a pretty nice performance improvement, but maybe copying 6'000'000 objects would negate the performance improvements.

Reading the source some more, the first branch would return a temporary and RVO there is guaranteed. Considering that the size of the vector "blueprints" is constant (3 pointers), it might work better than I expect.

Reply:
Even if no RVO is involved the result should be moved, so no performance hit, right?

Reply:
You're right. It will definitely be possible to move the vector, instead of copying.

Reply:
I'm not really good at asm reading, I'll just live link here: https://godbolt.org/z/9nhKPc

Reply:
Good point. But there must be a reason this wasn't done initially, no?

The allocation will happen anyway in the caster, then, so this could be slower if the move-constructor is costly?

Reply:
(For the record, I already wanted to complain about `std::unique_ptr<Vector>(new Vector());`, but then `make_unique` is only C++14 :-( )

Reply:
`make_unique` is just syntax sugar, no? In comparison, `make_shared` can IIRC avoid a memory allocation compared to the naive initialization.

Reply:
> `make_unique` is just syntax sugar, no? In comparison, `make_shared` can IIRC avoid a memory allocation compared to the naive initialization.

I think it is, yes. But it's shorter (doesn't repeat the type) and makes things feel more atomic (whereas `unique_ptr<T>(new T())` allows you to be tempted to split up the `new` and the construction of `unique_ptr`, which could result in leaks on exceptions, etc).

But it's C++14, so nevermind all that :-D

Reply:
Since I was the one bringing up RVO and whether it works...

```asm
; GCC
.L32:
        mov     rax, QWORD PTR [rsp]
        mov     QWORD PTR [r13+0], rax
        mov     rax, QWORD PTR [rsp+8]
        mov     QWORD PTR [r13+8], rax
        mov     rax, QWORD PTR [rsp+16]
        mov     QWORD PTR [r13+16], rax
.L24:
        mov     rax, r13
```

In other words, after populating the vector, we move 3 words, from the top of the stack into `R13` and return a little later. This is what a move looks like. At O3 it will be further optimized to 128bit move + 64bit move, instead of 3 x 64bit moves.

```asm
; Clang
        mov     rax, r14
```

You guessed it. RVO works here.

```asm
; MSVC
$LN3@to_vector:
        mov     rax, QWORD PTR vec$1[rsp]
        mov     QWORD PTR [rsi], rax
        mov     QWORD PTR [rsi+8], rcx
        mov     rax, QWORD PTR vec$1[rsp+16]
        mov     QWORD PTR [rsi+16], rax
$LN36@to_vector:
        mov     rax, rsi
```

This is the exact same thing as GCC, just with different registers. Except that I couldn't get MSVC to use XMM0 and perform a 128bit move.

&nbsp;

The branch with the temporary gets optimized to a `memcpy` and RVO is used. The assembly is just beautiful in this case.

&nbsp;

@YannickJadoul 

> so this could be slower if the move-constructor is costly

True, but we're talking about a `std::vector`-like container. You're not moving the `T`s. You're moving the container "blueprints". Is there a reasonable "vector-like" container with expensive move constructor?

Reply:
Whooooa, great investigation, @bstaletic! :-D

> True, but we're talking about a `std::vector`-like container. You're not moving the `T`s. You're moving the container "blueprints". Is there a reasonable "vector-like" container with expensive move constructor?

Probably not; I'm just thinking of some custom thing, but then it might be the user's own problem if they can't implement a decent move constructor.
On the other hand, I'm wondering that if it doesn't matter, why not just keep it like this?

Reply:
> On the other hand, I'm wondering that if it doesn't matter, why not just keep it like this?

Because readability and maintainability. It's easier to reason about the code without needless indirections.